rfc: extend updates rfc with constraints section (#1001)
edgelesssys/constellation: rfc/updates.md

  To learn if microserviceVersion A.B.C is compatible with Kubernetes version X.Y, one has to check whether the Constellation version is compatible with the Kubernetes version X.Y.
- Each Constellation version X.Y.Z is compatible with all patch versions of the next Constellation version X.Y+1.
- The individual versions of Microservice, image and CLI all have to be compatible with the targeted Kubernetes version.
  Each component can be upgraded independently.
  This means that during a Kubernetes upgrade the oldest Constellation component has to be compatible with the new Kubernetes version.
  If not, this oldest component has to be updated first.
- For Kubernetes versions the version-target of an upgrade has to be supported by the current Constellation version.
  The currently running Kubernetes version is not relevant for the Kubernetes upgrade.

These constraints are enforced by the CLI.
If users decide to change specific versions by changing the Kubernetes resources, nothing is stopping them.

The compatibility information should be separated from the enforcement code.

> This way a minimal implemention can be created where the compatibility information is embedded into the CLI.

Reviewer: ```suggestion
This way a minimal implementation can be created where the compatibility information is embedded into the CLI.
```

Warning: Please backup any important components before upgrading Kubernetes
Apply change [yes/No]?
```

# Compatibility

> `constellation-upgrade` has to handle the version of four components: CLI, image, microservices and Kubernetes.

Reviewer: ```suggestion
`constellation upgrade` has to handle the version of four components: CLI, image, microservices and Kubernetes.
```
This is not a new CLI binary, right?

To do this correctly and keep the cluster in a working condition some constraints are required.

> - Microservice, image and CLI are versioned in lockstep, i.e. each release a new version of all componentes is released.

Reviewer: ```suggestion
- Constellation microservices, VM image and the CLI are versioned in lockstep, i.e. each release a new version of all components is released.
```

>   Each of these version numbers is one Constellation version.

Reviewer: I think this is somewhat hard to understand. Maybe more clear: 
`The Constellation version references all of these components`.
Still, if we want to say that each component has the same semver attached to it, we already do this in the sentence before.

> - Each Constellation version is compatible to three or four Kubernetes versions.

Reviewer: ```suggestion
- Each Constellation version is compatible with three or four Kubernetes versions.
```

>   When a new Kubernetes version is released Constellation will support four versions for one release cycle; before phasing out the oldest Kubernetes version.

Reviewer: ```suggestion
  When a new Kubernetes version is released, Constellation will support four versions for one release cycle; before phasing out the oldest Kubernetes version.
```

  To learn if microserviceVersion A.B.C is compatible with Kubernetes version X.Y, one has to check whether the Constellation version is compatible with the Kubernetes version X.Y.

>   The Constellation version will be the same as the microserviceVersion: A.B.C. Same goes for the image and CLI.

Reviewer: We've already established this a few points before.

Author: Can't find where we explicitly state it like that. Doesn't hurt to repeat it?

Reviewer: I think we already somewhat mention this in the first point:
> Constellation microservices, VM image and the CLI are versioned in lockstep, i.e. each release a new version of all components is released.
  The Constellation version references all of these components.

If we want to also mention the semver, I think we should do this there. If needed we can then say something like `since the microserviceVersion is the same as the CLI version ...`.

> - Each Constellation version X.Y.Z is compatible with all patch versions of another Constellation version X.Y+1.

Reviewer: ```suggestion
- Each Constellation version X.Y.Z is compatible with all patch versions of the next Constellation minor version X.Y+1.
```

- The individual versions of Microservice, image and CLI all have to be compatible with the targeted Kubernetes version.

>   Without loss of generality, this means that during a Kubernetes upgrade the oldest Constellation component has to be compatible with the new Kubernetes version.

Reviewer: ```suggestion
This means that during a Kubernetes upgrade the oldest Constellation component has to be compatible with the new Kubernetes version.
```
We should probably explicitly establish that one can upgrade e.g. only the microservices. Then it is clear what "**oldest** Constellation component" means. If this is what you are going for here.

  If not, this oldest component has to be updated first.
- For Kubernetes versions the version-target of an upgrade has to be supported by the current Constellation version.
  The currently running Kubernetes version is not relevant for the Kubernetes upgrade.

These constraints are enforced by the CLI.

> If users decide to change specific versions on their own nothing is stopping them.

Reviewer: ```suggestion
If users decide to change specific versions by changing the Kubernetes resources itself, nothing is stopping them.
```
